fedopt_guide/cifar10/federated_cifar10.py

```diff
@@ -0,0 +1,136 @@
+  The iterative process must also have a callable attribute `get_model_weights`
+  that takes as input the state of the iterative process, and returns a
+  `tff.learning.ModelWeights` object.
+  Args:
+    iterative_process_builder: A function that accepts a no-arg `model_fn`, and
+      returns a `tff.templates.IterativeProcess`. The `model_fn` must return a
+      `tff.learning.Model`.
+    client_epochs_per_round: An integer representing the number of epochs of
+      training performed per client in each training round.
+    client_batch_size: An integer representing the batch size used on clients.
+    clients_per_round: An integer representing the number of clients
+      participating in each round.
+    client_datasets_random_seed: An optional int used to seed which clients are
+      sampled at each round. If `None`, no seed is used.
+    crop_size: An optional integer representing the resulting size of input
+      images after preprocessing.
+    total_rounds: The number of federated training rounds.
+    experiment_name: The name of the experiment being run. This will be appended
+      to the `root_output_dir` for purposes of writing outputs.
+    root_output_dir: The name of the root output directory for writing
+      experiment outputs.
+    **kwargs: Additional arguments configuring the training loop. For details
+      on supported arguments, see
+      `federated_research/utils/training_utils.py`.
+  """
+
+  crop_shape = (crop_size, crop_size, 3)
+
+  cifar_train, _ = cifar10_dataset.get_federated_datasets(
+      train_client_epochs_per_round=client_epochs_per_round,
+      train_client_batch_size=client_batch_size,
+      crop_shape=crop_shape)
+
+  _, cifar_test = cifar10_dataset.get_centralized_datasets(
+      train_batch_size=client_batch_size,
+      crop_shape=crop_shape)
+
+  input_spec = cifar_train.create_tf_dataset_for_client(
+      cifar_train.client_ids[0]).element_spec
+
+  model_builder = functools.partial(
+      resnet_models.create_resnet18,
+      input_shape=crop_shape,
+      num_classes=NUM_CLASSES)
+
+  loss_builder = tf.keras.losses.SparseCategoricalCrossentropy
+  metrics_builder = lambda: [tf.keras.metrics.SparseCategoricalAccuracy()]
+
+  def tff_model_fn() -> tff.learning.Model:
+    return tff.learning.from_keras_model(
+        keras_model=model_builder(),
+        input_spec=input_spec,
+        loss=loss_builder(),
+        metrics=metrics_builder())
+
+  training_process = iterative_process_builder(tff_model_fn)
+
+  client_datasets_fn = training_utils.build_client_datasets_fn(
+      dataset=cifar_train,
+      clients_per_round=clients_per_round,
+      random_seed=client_datasets_random_seed)
+
+  test_fn = training_utils.build_centralized_evaluate_fn(
+      eval_dataset=cifar_test,
+      model_builder=model_builder,
+      loss_builder=loss_builder,
+      metrics_builder=metrics_builder)
+
+  validation_fn = lambda model_weights, round_num: test_fn(model_weights)
```

Comment:
A necessary follow-up would be to allow a federated evaluation instead (depending on what the user chooses) which would allow us to aggregate client-level evaluation statistics (eg. minimum accuracy across clients).

Reply:
In this case would the test data also be distributed in the same non iid manner across clients?